src/services/refactors/extractType.ts

```diff
@@ -147,6 +147,11 @@ namespace ts.refactor {
                     const symbol = checker.resolveName(node.typeName.text, node.typeName, SymbolFlags.TypeParameter, /* excludeGlobals */ true);
                     const declaration = tryCast(symbol?.declarations?.[0], isTypeParameterDeclaration);
                     if (declaration) {
+                        if (declaration.name.escapedText === node.typeName.escapedText && rangeContainsSkipTrivia(declaration, selection, file)) {

```

Comment:
Are we always certain that the files are the same between the use-site and the declaration?

Reply:
`resolveName` doesn’t resolve aliases (well, it does in order to check that the target matches the flags you asked for, but it doesn’t return the target), so yes.

Reply:
Well, a better answer is the fact that it’s a type parameter

Reply:
> Well a better answer is the fact that it's a type parameter.

Boy, do I have a surprise for you!

```ts
// @Filename: a.ts
interface Foo<T extends { prop: T }> {}

// @Filename: b.ts
interface Foo<T extends { prop: T }> {}
```

Reply:
does it mean instead of checking the first declaration only (`symbol?.declarations?.[0]`), we need to check all of them and see if any falls into the same range?

would be nice to at least add that as a test case.

Reply:
> Boy, do I have a surprise for you!

@DanielRosenwasser I've added a test for this case.

Reply:
![homer receding into hedges](https://c.tenor.com/pXe-GokoyA0AAAAC/the-s-impsons-bush.gif)

Reply:
Good start with that test case, but I think there's a lurking bug.

Reply:
I hate to drag this out longer, but does this work for if the same file has multiple declarations too?

```ts
interface Foo<T extends { prop: T }> {}
interface Foo<T extends { prop: [|T|] }> {}
```

Seems like you want something along the lines of

```ts
declarations.some(declaration => isTypeParameterDeclaration(d) && declaration.name.escapedText === node.typeName.escapedText && d.getSourceFile() === file && rangeContainsSkipTrivia(declaration, selection, file))
```

as your check

Reply:
Ouch! Will add changes.

![loop](https://user-images.githubusercontent.com/509265/151513679-7fbcacbf-de40-4826-8760-f10e024d2a02.gif)